swift: split build & Darwin support.

stephank opened the pull request:
###### Description of changes

This PR splits up the Swift build into smaller derivations, and brings up Darwin support for the toolchain. Fixes #37473.

I've kept the same maintainers listed as before, also for the new derivations that were already part of the toolchain. Hopefully that's okay, otherwise, let me know.

To perform the split, I've added a `swiftPackages`. It's not very large, but avoids cluttering the global namespace, and helps set up a scope with the required clang, macOS SDK, etc.

Tests are a pain point, and I've disabled them in many places. I believe we can fix many of the tests, but it'll take more effort.

### Usage

The `swift` attribute now gets you just the Swift compiler (`swift` and `swiftc`). These tools are now wrapped, and a new setup-hook adds `lib/swift` of inputs to the Swift include path automatically. This is especially important on Darwin, where we separately package system frameworks.

The attribute `swiftpm` installs `swift-build`, `swift-test`, `swift-run`, though these are usually invoked without the hyphen by users (`swift build`). Additionally, a new setup-hook provides default build and check phases for convenience.

### Build process

Previously, our Swift build used the Python-based 'build-script' to build Swift, which I believe matches official Swift CI. This new build bypasses the Python scripts and invokes CMake directly. I personally find this approach better for us, because:

- We can split up the build. Smaller derivations = shorter iteration times and more granular caching. Especially for bringing up Darwin support, I've found quicker builds to be a great help.
- I reached a point where the `build-presets.ini` patches were becoming hard to stack and maintain.
- CMake scripts already have sane defaults for many options.
- We can leverage existing Nix CMake smarts.

The basic compiler derivation now builds primarily CMark, Clang, and Swift. We also build LLDB, but only because the REPL uses it. On Darwin, there is also some extra work for back-deployment, though I've not actually tested if/how back-deployment works.

On top of this, I implemented a wrapper similar to cc-wrapper.

I've separately packaged Dispatch, Foundation and XCTest. Dispatch is based on an existing derivation, but Swift support was added.

The second complex build is for SwiftPM. We match the official build process here in first doing a CMake build, then using that to build SwiftPM itself. I wrote a small swiftpm2nix tool based on shell + jq, which actually replaces the bulk of the `fetchgit` source derivations from the old Swift build.

Beyond that, I also packaged the other parts of the SDK, but believe only sourcekit-lsp is currently useful.

### Darwin stdenv

The biggest change to stdenv is that Swift modules are installed. These are `os` and `Darwin` for libsystem, and per-framework Swift overlays where available. I do have a small list of modules from the SDK that I couldn't place yet:

- AppleArchive
- Compression
- DataDetection
- ExtensionFoundation
- ExtensionKit
- QuickLookUI
- ShazamKit
- simd

I skipped these to limit time spent polishing details, and hope we can fix these as they become an issue.

Maybe it's possible to avoid these changes to existing SDK derivations by having separate derivations for all the Swift-specific stuff, but I haven't tried this, and wonder if it's worth it. Apple seems to treat Swift more and more as a core part of the system, and it'd make packaging a bit more tedious.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

stephank committed:
swift: compiler only build & darwin support

stephank committed:
swiftPackages.Dispatch: 5.5 -> 5.7

stephank committed:
swiftPackages.Foundation: init at 5.7

stephank committed:
swiftPackages.XCTest: init at 5.7

stephank committed:
swiftpm: init at 5.7

stephank committed:
swiftpm2nix: init

stephank committed:
sourcekit-lsp: init at 5.7

stephank committed:
apple_sdk_11_0: install swift modules

stephank committed:
apple_sdk_11_0: Foundation depends on Combine

stephank committed:
swift-docc: init at 5.7

stephank committed:
swift-driver: init at 5.7

Atemu commented on pkgs/development/compilers/llvm/11/compiler-rt/default.nix:
```diff
@@ -59,8 +60,9 @@ stdenv.mkDerivation {
-  ] ++ lib.optional stdenv.hostPlatform.isAarch32 ./armv7l.patch;
-
+    ../../common/compiler-rt/darwin-plistbuddy-workaround.patch
+    ./armv7l.patch
```
Why is this patch now here unconditionally?

stephank replied:
See: https://github.com/NixOS/nixpkgs/pull/186575#pullrequestreview-1074122694

Atemu commented on pkgs/development/compilers/llvm/11/compiler-rt/default.nix:
```diff
@@ -59,8 +60,9 @@ stdenv.mkDerivation {
     ../../common/compiler-rt/libsanitizer-no-cyclades-11.patch
-  ] ++ lib.optional stdenv.hostPlatform.isAarch32 ./armv7l.patch;
-
+    ../../common/compiler-rt/darwin-plistbuddy-workaround.patch
```
Darwin-only?

Atemu commented on pkgs/development/compilers/llvm/13/compiler-rt/default.nix:
```diff
@@ -62,9 +63,11 @@ stdenv.mkDerivation {
-    ++ lib.optional stdenv.hostPlatform.isDarwin ./darwin-targetconditionals.patch
-    ++ lib.optional stdenv.hostPlatform.isAarch32 ./armv7l.patch;
+    # Prevent a compilation error on darwin
+    ./darwin-targetconditionals.patch
```
Also darwin-only.

Atemu commented on pkgs/development/compilers/swift/compiler/default.nix:
```diff
@@ -0,0 +1,555 @@
+      #appletvsimulator
+      #watchos
+      #watchsimulator
+    }.${targetPlatform.darwinPlatform}
```
Not 100% necessary buy a default via `or` would be nice to have.

stephank replied:
Hmm, I'm not sure that'd even build. Isn't an error at evaluation time better in this case?

It looks like we do have some iOS target support, but I didn't actually test it. Other than that, this set is already broader than what we support in Nixpkgs.

Atemu replied:
`abort` with a helpful message also counts as a default in my books ;)

An attr not found exception cannot be caught in any way and isn't very user-friendly.

stephank replied:
I added:

```
or (throw "Cannot build Swift for target Darwin platform '${targetPlatform.darwinPlatform}'")
```

Atemu commented on pkgs/development/compilers/swift/sourcekit-lsp/default.nix:
```diff
@@ -0,0 +1,74 @@
+  # Generated by swiftpm2nix.
+  generated = callPackage ./generated { };
+
+  # On Darwin, we only want ncurses in the linker search path, because headers
+  # are part of libsystem. Adding its headers to the search path causes strange
+  # mixing and errors.
+  # TODO: Find a better way to prevent this conflict.
+  ncursesInput = if stdenv.isDarwin then ncurses.out else ncurses;
```
Is this a general problem on Darwin or only for swift packages?

stephank replied:
I think this may be specific to the LLVM modules logic, because if it was broader, I feel like we would've encountered it elsewhere. If so, it's possible ObjC can also cause this (maybe even C? I believe the flag is `-fmodules`.)

Maybe we should eventually get rid of some of these headers in libsystem. It currently ships headers for ncurses, sqlite, krb5, maybe some others, but we package those separately (and libsystem doesn't even provide implementations of those). A similar discussion came up in: https://github.com/NixOS/nixpkgs/pull/186251#issuecomment-1213605428

stephank commented:
cc swift maintainers: @dtzWill @trepetti @dduan @Trundle
cc swift-corelibs-libdispatch maintainer: @cmm
cc for darwin stdenv: @toonn? Should we ping more?

stephank commented:
While trying to build mpv, I noticed it invokes the compiler as `swift -frontend`, which was not wrapped. I modified the wrapper to also handle that specific case.

The build for mpv then fails with the same CoreVideo error as my little SwiftUI test app. Still debugging that issue.

stephank committed:
apple_sdk_11_0: Add SwiftUI dependencies

stephank committed:
apple_sdk_11_0: Fix missing CoreVideo include

stephank commented:
Some more tweaks:

- The CoreVideo issue appears to be an actual SDK issue. Should've checked this earlier, but the haeder file on my 12.3 SDK _does_ include `stdint.h`, while the 11.1 SDK we use doesn't. I've added a patch for it in https://github.com/NixOS/nixpkgs/pull/189977/commits/ca8576437f1faf887d57bb922ad9a7193a20efed.

- SwiftPM now propagates Foundation as a native input. It uses Foundation on the host platform while building package manifests. (Even though the target platform might use a different Foundation. Haven't tested this scenario, really, it might pop up when targeting iOS.)

- I moved around the `swift -frontend` wrapper fix, because it was in a part of the wrapper with a modified PATH. This caused the swift frontend to fail translating e.g. `swift build` to `swift-build`.

- I replaced an external patch for LLVM I had included as a file with a `fetchpatch` invocation.

- Added dependencies for the SwiftUI system framework.

I can now build a SwiftUI app! Will try mpv next, but I also rebased on current staging, and need to do a large rebuild.

stephank commented:
- mpv builds, with only small changes: https://github.com/stephank/nixpkgs/commit/e4566d271ccbd2d94fbc78bb39f66de9184d5eae. Not part of this PR, think it's better to do it separately later?
- SwiftPM now propagates `vtool` as a native input. This was the last remaining tool that was causing prompts to installed Command-Line Tools when using SwiftPM in a nix shell.
- I've been testing the recent changes on aarch64-darwin, and have kicked off a build on x86_64-darwin to verify.

toonn commented:
Re testing mpv, if it is not built with Swift it should spit out a warning:

```
[vo/gpu] opengl cocoa backend is deprecated, use vo=libmpv instead
```

It will also not be able to go into fullscreen mode when you press `f` (default shortcut).

This is probably not a thorough test of the functionality Swift adds to mpv but may be sufficient?

Atemu commented:
You're not actually using the new swift UI in that case. You need to add `--vo=libmpv` and then it works beautifully.

Building with `"--disable-macos-cocoa-cb"` makes it use libmpv by default but I'm not sure that's the best way of achieving that.

toonn commented on pkgs/os-specific/darwin/apple-sdk-11.0/apple_sdk.nix:
```diff
@@ -163,6 +171,15 @@ in rec {
+    CoreVideo = lib.overrideDerivation super.CoreVideo (drv: {
+      installPhase = drv.installPhase + ''
+        # When used as a module, complains about a missing import for
+        # Darwin.C.stdint. Apparently fix in later SDKs.
```
```suggestion
        # Darwin.C.stdint. Apparently fixed in later SDKs.
```

toonn reviewed (commented):
The Darwin stdenv changes look OK to me. Though I'm not so familiar with aarch64-darwin, maybe @thefloweringash can take a look as well?

toonn commented:
@Atemu, this is a way to test mpv is in fact using the Swift built libmpv. I'm saying if it was not then you'd get that error and fullscreen wouldn't work. This was just advice in response to stephank's edit:

> (EDIT: Also not sure what functionality to actually check at run-time. Never used mpv before myself.)

stephank commented:
Thanks for taking a look. Mpv fullscreen does work in the Swift enabled build (and not with my regular nix-darwin install), so that's a good sign.

For the REPL, it appears the logic is in swift-driver. The `swift` and `swiftc` executables forward calls to `swift-driver` if it exists. Current idea is to add swift-driver to the installation via the wrapper, and this works, but the new driver is not forwarding linker flags correctly. Once that's fixed, I also need to add the LLDB build somewhere

tjni commented:
Would you mind if I extracted [this commit](https://github.com/NixOS/nixpkgs/pull/189977/commits/762b2647f4092ae3e42ba2c59527cf58fa90425f) and [this commit](https://github.com/NixOS/nixpkgs/pull/189977/commits/a62cba241876ab2fd30d1bd9a536c5e65b21c812) into a separate PR to merge before this is ready? On my aarch64-darwin machine, they resolve building [cargo-watch](https://github.com/NixOS/nixpkgs/issues/146349) and [xcbuild](https://github.com/NixOS/nixpkgs/pull/161561#issuecomment-1056063573).

stephank commented:
> Would you mind if I extracted [this commit](https://github.com/NixOS/nixpkgs/pull/189977/commits/762b2647f4092ae3e42ba2c59527cf58fa90425f) and [this commit](https://github.com/NixOS/nixpkgs/pull/189977/commits/a62cba241876ab2fd30d1bd9a536c5e65b21c812) into a separate PR to merge before this is ready? On my aarch64-darwin machine, they resolve building [cargo-watch](https://github.com/NixOS/nixpkgs/issues/146349) and [xcbuild](https://github.com/NixOS/nixpkgs/pull/161561#issuecomment-1056063573).

Please do! I had no idea this was broader than Swift. 🙂

stephank committed:
xcbuild: add JSON variant of SDKSettings

stephank committed:
xcbuild: exit with error if --find fails

stephank commented:
Rebased an updated to Swift 5.7.

This highlighted some issues with static linking that are now fixed. I also install clang in swift-unwrapped, because swift now uses some (optional) functionality not in upstream (index while building).

The new swift-driver is working and the default on this branch, but I still need to add the lldb buiild for the repl.

stephank commented:
The REPL now works, though it prints the following warnings on startup, and I don't understand why:

```
warning: (arm64) /nix/store/b7srw4l9v7zh38h15xr5axh0a6pai3yv-swift-5.7-lib/bin/repl_swift(0x0000000100000000) address 0x0000000100000000 maps to more than one section: repl_swift.__TEXT and repl_swift.__TEXT
warning: (arm64) /nix/store/b7srw4l9v7zh38h15xr5axh0a6pai3yv-swift-5.7-lib/bin/repl_swift(0x0000000100000000) address 0x0000000100008000 maps to more than one section: repl_swift.__DATA_CONST and repl_swift.__DATA_CONST
warning: (arm64) /nix/store/b7srw4l9v7zh38h15xr5axh0a6pai3yv-swift-5.7-lib/bin/repl_swift(0x0000000100000000) address 0x000000010000c000 maps to more than one section: repl_swift.__DATA and repl_swift.__DATA
```

It also looks like LLDB has the ability to attach a REPL to a running process? I'm not sure how this works, and if it's available from the command-line. I've currently chosen to hide LLDB (and Clang) for Swift 'internal' use only, so as to not clobber PATH with them.

I currently tested only `aarch64-darwin`, and am still running builds for the other three platforms.

I think this is a good point for me to stop tweaking this PR. If it's okay, I'd like to focus on blocking issues, review feedback, and moving this forward with the functionality as-is.

stephank commented:
Introducing Clang in the install broke the Linux build, because it was using that (unwrapped) Clang as a linker. This is now fixed.

At some point, I thought the wrapper was causing this, so I also sat down and dug into the CLI parsing code in Swift. I updated the wrapper to match that logic, and it's now in much better shape.

The branch is rebased on latest staging, and verified again on `x86_64-darwin`, `aarch64-darwin`, and `aarch64-linux`. I'm running an `x86_64-linux` build on nixbuild.net now.

trepetti commented on pkgs/development/compilers/swift/swift-driver/default.nix:
```diff
@@ -0,0 +1,81 @@
+    ./patches/linux-fix-linking.patch
+    (fetchpatch {
+      url = "https://patch-diff.githubusercontent.com/raw/apple/swift-driver/pull/1197.patch";
+      hash = "sha256-eVBaKN6uzj48ZnHtwGV0k5ChKjak1tDCyE+wTdyGq2c=";
```
This patch has a different SHA now:
> `specified: sha256-eVBaKN6uzj48ZnHtwGV0k5ChKjak1tDCyE+wTdyGq2c=`
> `got:          sha256-NeYu7Ex2F5AgclSP5Y7z507ZxdVMC9XPUlkY64nJIG4=`
Are were sure that this asset is stable?

stephank replied:
Ah, darn, this is something I'm trying to upstream: https://github.com/apple/swift-driver/pull/1197

I suppose it's not stable until it is merged. I can copy it to nixpkgs, or we can wait a bit? (I'm hoping they'll merge it soon.)

trepetti replied:
Oh that's great. Your call. I'm just trying to do a shakedown on Linux. I don't have the compute resources for the full closure of this PR, but as one of the Linux maintainers thought I should make sure that side was covered. Thanks for this huge contribution!

tjni replied:
Patch can be stable if referring to individual commit without the /pull/ part of the URL. For example:

https://github.com/apple/swift-driver/commit/d3ef9cdf4871a58eddec7ff0e28fe611130da3f9.patch

It is better when available to reference the merge commit so that it’ll be associated with a branch, but that commit will always be there even if the PR changes.

stephank replied:
I rebased and used the commit URL. I left a TODO comment; but hope it's otherwise not a blocker?

stephank commented:
I noticed builds on Linux had the Swift compiler in rpath, and fixed it. This fix also produces paths to `$lib` in the output of `swift -frontend -print-target-info`. On Linux, `$out` is now ~800MB, while `$lib` is ~200MB.

There are more gains to be made in the `$lib` closure, but I hope we can fix those later, and not block this PR too long.

I pushed binaries of a `x86_64-linux` build of a3cc6c1 to Cachix, to hopefully help review:
- https://app.cachix.org/cache/stephank
- Swift: /nix/store/fcidx377livn242mayq0f1w5nz6c0kyz-swift-wrapper-5.7
- SwiftPM: /nix/store/dv2yf9na48ab0gbwc06n27hay1jh67ff-swiftpm-5.7
- SourceKit-LSP: /nix/store/fgspwy6wjfrcb7hvs9vwarx0n3gnxwnl-sourcekit-lsp-5.7

stephank commented:
Binaries for a3cc6c1 on `aarch64-darwin`:

- https://app.cachix.org/cache/stephank
- Swift: /nix/store/w70gh1qmz9c77lqdlir51wdmlib3h3jx-swift-wrapper-5.7
- SwiftPM: /nix/store/yffdw3nccrvx0a3j6fgv0f6im5c1z9h5-swiftpm-5.7
- SourceKit-LSP: /nix/store/6q0vaixz3fqhfy07r0hhscbz8bqqdqy2-sourcekit-lsp-5.7

stephank commented:
Binaries for a3cc6c1 on `x86_64-darwin`:

- https://app.cachix.org/cache/stephank
- Swift: /nix/store/3mdv0qc7gmv0m2dx07kbrj3rxsxq8akk-swift-wrapper-5.7
- SwiftPM: /nix/store/xfrnscciqlx0mcfnv6s3869x4mgpm0rk-swiftpm-5.7
- SourceKit-LSP: /nix/store/xznwh25xdwnqq82d7s0wc20k5r4kcg5n-sourcekit-lsp-5.7

Atemu commented:
I think it would.

stephank commented:
I rebased the PR.

Binaries for de58d88 on `x86_64-linux`:
- https://app.cachix.org/cache/stephank
- Swift: /nix/store/j5x1dabcaf8f3isz6n5z8fy51wl0wdwp-swift-wrapper-5.7
- SwiftPM: /nix/store/42s94d4f555ynns5y3ld1mhickbvxh3i-swiftpm-5.7
- SourceKit-LSP: /nix/store/cnjkq301r1gaj778jr7wymx7dpcw3x80-sourcekit-lsp-5.7

trepetti reviewed (approved):
Hi Stephan, testing on the Linux side of things has gone great. Thank you for this awesome contribution.

stephank commented:
I rebased and fixed a conflict with #181764. I don't believe that PR affects Darwin. On Linux, the fix from that PR doesn't appear to be necessary anymore.

Binaries for d63aef6c479021ebc0bcb67ea852bee0041d4474 on `x86_64-linux`:
- https://app.cachix.org/cache/stephank
- Swift: /nix/store/7yk9wjj2zjbdm33qn2j2qpjvm7p1y0zd-swift-wrapper-5.7
- SwiftPM: /nix/store/s976zxa2y0gh04hkgcdjkvs4h51927qa-swiftpm-5.7
- SourceKit-LSP: /nix/store/1q8534vz6mr1p3xh4dhd6fk5zc2335n2-sourcekit-lsp-5.7

toonn reviewed (commented):
Wow, this work is considerable to say the least. Makes reviewing quite a task.
I'd like to ask to avoid further rebasing while reviews are on-going. Fixup commits are a good option for an autosquash rebase afterwards.

Thank you for the effort that must've gone into this! ❤️

stephank commented on pkgs/development/compilers/swift/compiler/patches/swift-nix-resource-root.patch:
```diff
@@ -0,0 +1,67 @@
++  // prefer detecting the Clang resource root via NIX_CC, to allow swapping in
++  // a different stdenv. However, always honor a user-provided `-resource-dir`.
++  auto nixCC = llvm::sys::Process::GetEnv("NIX_CC");
++  if (nixCC && !Args.hasArgNoClaim(options::OPT_resource_dir)) {
```
I couldn't find a way for ClangImporter, unfortunately. I don't believe there is a way to access the frontend arguments there. Probably because ClangImporter is used outside the frontend as well.

stephank commented on pkgs/development/compilers/swift/libdispatch/disable-swift-overlay.patch:
```diff
@@ -0,0 +1,35 @@
+Enabling Swift support is normally intended for building an overlay for a
```
`ENABLE_SWIFT` simply enables Swift support. The trouble is that it messes up the installation compared to a build with no Swift support for some reason. This patch fixes that so the same build can be used with and without Swift.

stephank commented on pkgs/development/compilers/swift/sourcekit-lsp/default.nix:
```diff
@@ -0,0 +1,74 @@
+    homepage = "https://github.com/apple/sourcekit-lsp";
+    platforms = with lib.platforms; linux ++ darwin;
+    license = lib.licenses.asl20;
+    maintainers = with lib.maintainers; [ dtzWill trepetti dduan trundle stephank ];
```
I believe all of these were already part of the old build, just not separate derivations. I ping'd all maintainers in this PR, at least.

toonn replied:
Ah yes, it's just because it was split out into many expressions. I've seen reactions in the thread from everyone except for @dtzWill and @dduan. So I'm sure it's fine unless someone raises objections.

stephank commented on pkgs/development/compilers/swift/swift-driver/patches/disable-catalyst.patch:
```diff
@@ -0,0 +1,17 @@
+-      if self.canonicalName.hasPrefix("macosx") {
+-        self.versionMap = try keyedContainer.decode(VersionMap.self, forKey: .versionMap)
+-      } else {
+         self.versionMap = VersionMap()
```
That's right. I tried to avoid unnecessary whitespace changes in the diffs to keep them small.

stephank commented on pkgs/development/compilers/swift/swift-driver/patches/linux-fix-linking.patch:
```diff
@@ -0,0 +1,40 @@
+         parsedOptions.hasArgument(.enableExperimentalCxxInterop) ? .clangxx : .clang
+-      var clangPath = try getToolPath(clangTool)
++
++      // For Nix, prefer linking using the wrapped system clang, instead of using
```
They're entirely separate codebases, so I don't see an obvious way to do it.

The C++ code is kinda on the way out. We currently still need it to do the initial bootstrapping, but it's not used afterwards. In a future Swift release (not sure about the timeline on this), an existing Swift toolchain will be required to build Swift, so we'll need to start shipping some bootstrap binaries. I'm hoping the C++ side of things will simply go away at some point as a result.

stephank commented on pkgs/development/compilers/swift/swiftpm/patches/disable-sandbox.patch:
```diff
@@ -0,0 +1,21 @@
+     ) -> [String] {
+         #if os(macOS)
++        let env = ProcessInfo.processInfo.environment
++        if env["NIX_BUILD_TOP"] == nil || env["IN_NIX_SHELL"] != nil {
```
Nix uses its own sandboxing for builds, and that causes these calls to fail.

A Nix shell never has the Nix sandbox setup, I believe? Or are there situations where a Nix shell does have a sandbox?

stephank commented on pkgs/development/compilers/swift/swiftpm/patches/fix-clang-cxx.patch:
```diff
@@ -0,0 +1,121 @@
+Swiftpm may invoke clang, not clang++, to compile C++. Our cc-wrapper also
```
For swiftpm, it'd be quite simple, I think. We just need to detect `-std=` with one of the C++ variants.

I realize this patch isn't particularly small, but I was hesitant to modify cc-wrapper in this already large PR. It'd cause a big rebuild on Linux as well. So I prefer we do it separately.

stephank commented on pkgs/development/compilers/swift/swiftpm/setup-hook.sh:
```diff
@@ -0,0 +1,60 @@
+        $swiftpmFlags "${swiftpmFlagsArray[@]}"
+    )
+
+    swift-build --show-bin-path "${flagsArray[@]}"
```
I think this flag is specifically meant to be scriptable. `TERM=dumb` was to fix broken progress indicators in builds. It doesn't appear to be necessary here.

stephank commented on pkgs/development/compilers/swift/swiftpm/setup-hook.sh:
```diff
@@ -0,0 +1,60 @@
+
+# Helper used to find the binary output path.
+# Useful for performing the installPhase of swiftpm packages.
+swiftpmBinPath() {
```
I don't really like that this function is difficult to find, though. Pretty much every package using swiftpm will end up using this in their installPhase. I think we should add a Swift section to our documentation (when this lands on master), but I wonder if that's enough.

toonn replied:
I agree docs are a necessity. I'm not sure how we could make things any more discoverable though, hooks are particularly bad from that perspective in general.

stephank commented on pkgs/development/compilers/swift/wrapper/default.nix:
```diff
@@ -0,0 +1,59 @@
+
+    # This link is here because various tools (swiftpm) check for stdlib
+    # relative to the swift compiler. It's fine if this is for build-time
+    # stuff, but we should patch all cases were it would end up in an output.
```
Definitely a tricky issue. I don't know if this is limited to just the Swift toolchain, or broader.

One simple thing we can do is maybe add a check to one of the leaf packages like sourcekit-lsp, that it doesn't depend on `out`, but `lib`. That way we can catch any regressions in the toolchain, at least.

toonn replied:
A canary sounds like a good idea, especially if it doesn't add too much complexity.

stephank replied:
I found this was quite easy to do with just an attribute `disallowedRequisites`.

stephank commented on pkgs/development/compilers/swift/wrapper/wrapper.sh:
```diff
@@ -0,0 +1,291 @@
+fi
+
+PATH="$path_backup"
+# Old bash workaround, see above.
```
Hmm, I copied this from cc-wrapper: https://github.com/NixOS/nixpkgs/blob/1e624ddfc67e826862cd56b0b3c6bc4caea5e371/pkgs/build-support/cc-wrapper/cc-wrapper.sh#L220

The idea was to make the two diffable, but I'm not sure how well that holds up.

toonn replied:
Inherited weirdness is fine, I guess.

stephank commented on pkgs/development/tools/xcbuild/wrapper.nix:
```diff
@@ -73,7 +73,7 @@ while [ $# -gt 0 ]; do
          --find | -find | -f)
            shift
-           command -v $1 ;;
+           command -v $1 || exit 1 ;;
```
This is maybe because the `xcrun` script does not have `set -e`, but when the command is not found, it'd still 'exit 0' with no output. This was a problem somewhere (swiftpm or swift-driver, not sure from memory).

stephank commented on pkgs/os-specific/darwin/apple-sdk-11.0/default.nix:
```diff
@@ -24,16 +15,45 @@ let
 
-    passthru = {
-      inherit version;
+  CLTools_Executables = mkSusDerivation {
```
We only need CLTools_Executables for the ARC support library the Swift compiler links in. Swift requires SDK 10.15 at least, so there's no benefit doing it in the 10.12 SDK.

toonn replied:
Do the command-line tools need to stay in sync with the SDK? I'm not sure they shouldn't be an independent expression.

stephank replied:
Not necessary, but I'm not sure where else it'd belong. It can be a separate expression in a file in `pkgs/development/compilers/swift/compiler/`, but that seems out of place. Especially if we later find another use for it.

I was hoping we could also update these together and sourced from the same sucatalog, to prevent drift we later don't understand.

On the other hand, if we're going to use this as a template for newer SDKs, we'll have cruft in there that's not necessary right now. CLTools is only used from the SDK version the Swift compiler happens to use.

stephank commented on pkgs/development/compilers/swift/swiftpm/default.nix:
```diff
@@ -0,0 +1,469 @@
+    };
+    nativeBuildInputs = [ makeWrapper ];
+    # Required at run-time for the host platform to build package manifests.
+    propagatedBuildInputs = [ Foundation ];
```
Maybe this needs a better comment. It's swiftpm that requires it at run-time. So, a derivation that uses swiftpm to build something needs Foundation as a `nativeBuildInput`, and this makes that automatic when swiftpm is already a `nativeBuildInput`. It's needed for when swiftpm compiles `Package.swift`.

toonn replied:
Why would a run-time dependency of a tool need to be a build input to a dependent? Shouldn't run-time dependencies always be part of a tool's closure anyway?

stephank replied:
Foundation needs to be a nativeBuildInput of a package using swiftpm, because swiftpm performs a build step for `Package.swift` targeting the build platform. Having it the closure is not enough because only buildInputs are processed by setup hooks, which add the necessary search paths to compiler flags.

toonn replied:
OK, definitely confusing but I think I follow your reasoning.

stephank commented on pkgs/development/compilers/swift/swiftpm/default.nix:
```diff
@@ -0,0 +1,469 @@
+      + ''
+        # Builds often don't set a target, and our default minimum macOS deployment
+        # target on x86_64-darwin is too low. Harmless on non-Darwin.
+        export MACOSX_DEPLOYMENT_TARGET=10.15.4
```
Swift should be able to deploy all the way back to macOS 10.9, but I believe at least SDK 10.15 is required. I'm not sure if the toolchain itself runs on older than 10.15, but this failed to compile with a lower target.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-macos-monthly/12330/34

stephank commented on pkgs/development/compilers/swift/swiftpm/patches/llbuild-cmake-disable-rpath.patch:
```diff
@@ -0,0 +1,14 @@
+Specifying `-platform_version` targeting macos before 10.15 causes cctools ld
```
I believe so, yes, though I can no longer find the exact code in cctools that causes this behavior. I discovered this by testing various `ld` invocations.

stephank commented on pkgs/development/compilers/swift/wrapper/default.nix:
```diff
@@ -0,0 +1,59 @@
+, stdenv
+, swift
+, wrapperParams ? swift._wrapperParams
+, useSwiftDriver ? true, swift-driver
```
I think this is only useful if `swift-driver` is nullable, which in turn would only be useful if it was something a user has to explicitly inject in `callPackage`. In this case, `swift-driver` is always present, because it is already part of the swift scope.

stephank commented on pkgs/development/compilers/swift/wrapper/default.nix:
```diff
@@ -0,0 +1,59 @@
+, useSwiftDriver ? true, swift-driver
+}:
+
+stdenv.mkDerivation (wrapperParams // {
```
Good point. I think the reason it is an arg at all is because I was concerned about someone trying to use the wrapper with an overridden `swift` arg, where `swift._wrapperParams` may not be present and needs to be provided otherwise. Feels hacky, though.

It can be flipped if we handle `swiftDriver`, which is the only thing that depends on order. So we could maybe do:

```nix
stdenv.mkDerivation ({ ... } // wrapperParams // { swiftDriver = ... })
```

Though, thinking about it some more, if the intention is to use `wrapperParams` to override something in the derivation, I think there is no benefit over just using `overrideAttrs`?

toonn replied:
If we can drop it, let's drop it. Otherwise, I'm fine with the order.

stephank replied:
The arg was dropped.

stephank committed:
swiftpm2nix: reduce duplication in generated files

stephank committed:
swift: track version in a central sources.nix

stephank committed:
swiftPackages: update swiftpm2nix generated files

stephank committed:
Merge remote-tracking branch 'origin/staging' into HEAD

toonn committed:
swiftpm: improve setup hook shell conditions

stephank commented on pkgs/development/tools/swiftpm2nix/swiftpm2nix.sh:
```diff
@@ -0,0 +1,74 @@
+
+# Iterate dependencies and prefetch.
+sources=""
+jq -r '.object.dependencies[] | "\(.subpath) \(.packageRef.location) \(.state.checkoutState.revision)"' $stateFile \
```
As in `jq nix-prefetch-git`? I don't think that's necessary because nix provides those. Most scripts in nixpkgs use a nix-shell shebang, but in this case it's an installable (will work with `nix-shell -p swiftpm2nix`) wrapped to have those dependencies available.

toonn committed:
swiftpm: typo fixes

stephank committed:
swiftpm: factor out cmake glue installation

stephank committed:
swift: remove NIX_BUILD_TOP

stephank commented on pkgs/development/compilers/swift/libdispatch/default.nix:
```diff
@@ -0,0 +1,46 @@
+{ lib
```
That appears to be unused: https://github.com/NixOS/nixpkgs/blob/b2444828a1a6682a3976d8ab9758b26bca9a7592/pkgs/top-level/darwin-packages.nix#L197-L198

A separate attribute `swift-corelibs-libdispatch` already existed. That's this derivation. I simply moved it and enabled the Swift support.

toonn commented on pkgs/development/compilers/swift/foundation/default.nix:
```diff
@@ -0,0 +1,60 @@
+{ lib
```
I think we should really merge these but let's not hold this back on that. Maybe add a comment pointing to that package though?

stephank replied:
I added a TODO comment.

stephank committed:
swiftPackages.Foundation: add note about CF

stephank committed:
swift: remove internal wrapperParams arg

stephank committed:
swift: add canary to verify no compiler dependency

stephank commented on pkgs/development/compilers/swift/sourcekit-lsp/generated/default.nix:
```diff
@@ -0,0 +1,72 @@
+# This file was generated by swiftpm2nix.
```
I made some changes to swiftpm2nix to reduce the generated code. The `fetchgit` derivations are now partially derived from `workspace-state.json`, with only the hashes having to be filled in. The `Package.resolved` file is now fully derived from `workspace-state.json`.

stephank commented on pkgs/development/compilers/swift/sourcekit-lsp/default.nix:
```diff
@@ -0,0 +1,74 @@
+  pname = "sourcekit-lsp";
+
+  # Releases are made as part of the Swift toolchain, so versions should match.
+  version = "5.7";
```
I added a `sources.nix` that contains all fetch derviations, and all packages now refer to it.

stephank commented on pkgs/development/compilers/swift/compiler/default.nix:
```diff
@@ -0,0 +1,681 @@
+
+  preFixup = lib.optionalString stdenv.isLinux ''
+    # This is cheesy, but helps the patchelf hook remove /build from RPATH.
+    cd $NIX_BUILD_TOP
```
I combined the glue files in a `cmake-glue.nix`, and it also combines these shell snippets needed for installation.

stephank commented:
I did a merge with a recent(ish) staging (when I started working on it again), and was able to verify a build on `aarch64-darwin` with these changes.

Swift has since released 5.7.1, but we can do that separately.

stephank committed:
swiftPackages.Dispatch: fix linux build

ylh commented:
Reporting that this branch is of high enough quality that mpv, with some additional `buildInputs` tweaking, can now build with the Swift frontend instead of whatever `--disable-macos-cocoa-cb` gives us. This means nice things like better fullscreen support and the overlaid window decoration that vanishes along with the playback controls. Looking forward to this getting merged.

stephank commented:
Fwiw, I believe all feedback is addressed? Not sure what else I should do here. Would be great if we could move this forward!

toonn commented:
I've been waiting for another pair of eyes on this. If that's not forthcoming we can merge though IMO.

Atemu commented:
Even if there is some issue you overlooked, I think this is a major improvement over the status quo with no obvious flaws. Any further refinements can always be done afterwards.

If you looked over it and everything you found has been addressed, I think it's good enough for staging/unstable.

stephank commented:
Ping! Can we merge this? 🙂

toonn reviewed (commented):
Couple remaining questions.

toonn reviewed (approved):
LGTM
